In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report
)

In [2]:
# Load telecom KPI dataset from CSV file
data = pd.read_csv('telecom_kpi_data.csv')

# Display the data
print(data)

     rsrp_dbm  rsrq_db  sinr_db  cqi  tx_power_dbm  vswr  temperature_c  \
0      -101.0    -17.0     10.0    7          32.0   2.1           46.0   
1       -97.0    -15.0     10.0    6          34.0   2.3           47.0   
2       -95.0    -12.0      9.0    6          33.0   1.4           41.0   
3      -106.0    -15.0      4.0    4          31.0   2.2           52.0   
4       -84.0     -9.0     19.0   10          33.0   1.4           43.0   
..        ...      ...      ...  ...           ...   ...            ...   
495    -117.0    -20.0      7.0    5          31.0   2.7           54.0   
496    -109.0    -20.0     12.0    7          30.0   2.2           52.0   
497    -113.0    -18.0      6.0    5          26.0   2.4           51.0   
498     -81.0    -10.0     21.0   11          36.0   1.2           43.0   
499     -90.0    -14.0     15.0    8          34.0   1.5           42.0   

     prb_util_dl  drop_call_rate  ho_fail_rate  rrc_sr  cell_availability  \
0           54.0      

In [4]:
# Features and target
X = data.drop('faulty', axis=1)
y = data['faulty']
#print(X)
#print(y)

In [5]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

In [6]:
# Train classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [7]:
# Evaluation
print("Predictions:", y_pred)
print("Actual:", list(y_test))
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Predictions: [0 1 1 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 1 1
 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0
 0 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0
 0 0 0 0 0 0 1 0 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0
 0 0]
Actual: [0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
Accuracy: 0.9133333333333333
Precision: 0.8214285714285714
Recall: 0.7419354838709677
F1 Score: 0.7796610169491526

Confusion Matrix:
 [[114   5]
 [  8  23]]

Classification Report:
               precision    recall  f1-s


| | **Predicted NO** | **Predicted YES** |
|---|---|---|
| **Actual NO** | 114 (True Negatives) | 5 (False Positives) |
| **Actual YES** | 8 (False Negatives) | 23 (True Positives) |

---

## Key Metrics

### 1. Accuracy: 0.9133 (91.33%)

**Formula:**
$$Accuracy = \frac{TP + TN}{TP + TN + FP + FN} = \frac{23 + 114}{23 + 114 + 5 + 8} = \frac{137}{150} = 0.9133$$

**Interpretation:**
- The model correctly classifies **91.33%** of all cases
- Overall, 137 out of 150 predictions are correct
- **Limitation**: Can be misleading with imbalanced data

---

### 2. Precision: 0.8214 (82.14%)

**Formula:**
$$Precision = \frac{TP}{TP + FP} = \frac{23}{23 + 5} = \frac{23}{28} = 0.8214$$

**Interpretation:**
- When the model predicts a cell is faulty, it is correct **82.14%** of the time
- **23 out of 28** positive predictions are correct
- **Use case**: Important when the cost of false alarms is high

---

### 3. Recall (Sensitivity): 0.7419 (74.19%)

**Formula:**
$$Recall = \frac{TP}{TP + FN} = \frac{23}{23 + 8} = \frac{23}{31} = 0.7419$$

**Interpretation:**
- The model correctly identifies **74.19%** of actual faulty cells
- **23 out of 31** actual faults are detected
- **Use case**: Important when missing faults is costly

---

### 4. F1 Score: 0.7797 (77.97%)

**Formula:**
$$F1 = 2 \times \frac{Precision \times Recall}{Precision + Recall} = 2 \times \frac{0.8214 \times 0.7419}{0.8214 + 0.7419} = 0.7797$$

**Interpretation:**
- Harmonic mean of precision and recall
- Balances the trade-off between false positives and false negatives
- **77.97%** overall performance considering both metrics

---

## Classification Report Breakdown

### Class 0 (Non-faulty cells)
| Metric | Value | Interpretation |
|--------|-------|----------------|
| **Precision** | 0.93 | 93% of predicted non-faulty are actually non-faulty |
| **Recall** | 0.96 | 96% of actual non-faulty cells are correctly identified |
| **F1-score** | 0.95 | Excellent performance on normal cells |

### Class 1 (Faulty cells)
| Metric | Value | Interpretation |
|--------|-------|----------------|
| **Precision** | 0.82 | 82% of predicted faults are actually faulty |
| **Recall** | 0.74 | 74% of actual faults are detected |
| **F1-score** | 0.78 | Good performance on detecting faults |

---

## Performance Summary

### Averages
| Average Type | Precision | Recall | F1-Score |
|--------------|-----------|--------|----------|
| **Macro Avg** | 0.88 | 0.85 | 0.86 |
| **Weighted Avg** | 0.91 | 0.91 | 0.91 |

- **Macro Average**: Unweighted mean across classes (treats both classes equally)
- **Weighted Average**: Weighted by support (accounts for class imbalance)

---

## Key Insights

### 1. Strengths ✅
- **High Accuracy (91.33%)**: Model performs well overall
- **Excellent for Class 0 (Non-faulty)**: Precision (0.93) and Recall (0.96) are both high
- **Low False Positives**: Only 5 cases incorrectly flagged as faulty
- **Good Precision (0.82)**: Most flagged faults are real issues

### 2. Weaknesses ⚠️
- **Lower Recall for Faults (0.74)**: 8 faulty cells are missed (False Negatives)
- **Class Imbalance**: Only 31 faulty cells vs 119 non-faulty cells
- **F1 Score (0.78)**: Room for improvement in detecting faulty cells

### 3. Risk Assessment
| Scenario | Count | Risk |
|----------|-------|------|
| **False Positives** (Predicted faulty, actually normal) | 5 | Low - Unnecessary investigation |
| **False Negatives** (Predicted normal, actually faulty) | 8 | High - Undetected faults impacting service |

---

## Recommendations

1. **To Improve Recall**: 
   - Adjust decision threshold
   - Collect more fault data
   - Use SMOTE or oversampling techniques

2. **To Reduce False Negatives**: 
   - Use cost-sensitive learning
   - Add more fault-related features
   - Consider ensemble methods

3. **Model is Production-Ready** if:
   - Detecting faults is more important than avoiding false alarms
   - 74% recall is acceptable for your use case
   - Additional monitoring can catch missed faults

---

## Visual Performance Summary


In [8]:
# Feature importance
importance = pd.DataFrame({
    'Feature': X.columns,
    'Importance': model.feature_importances_
}).sort_values(by='Importance', ascending=False)

print("\nFeature Importance:\n", importance)


Feature Importance:
               Feature  Importance
9        ho_fail_rate    0.145436
10             rrc_sr    0.119982
0            rsrp_dbm    0.118002
12        packet_loss    0.107987
8      drop_call_rate    0.090310
11  cell_availability    0.083267
13         latency_ms    0.079789
5                vswr    0.070854
6       temperature_c    0.036304
2             sinr_db    0.032861
1             rsrq_db    0.032598
7         prb_util_dl    0.027022
4        tx_power_dbm    0.024928
3                 cqi    0.015649
14        alarm_count    0.015009
